File: src/main/kotlin/utils/RetryUtil.java
package utils;

import java.time.Duration;

public final class RetryUtil {
    private RetryUtil() {}

    public static void run(int attempts, Duration delay, Runnable action) {
        RuntimeException last = null;
        for (int i = 1; i <= attempts; i++) {
            try {
                action.run();
                return;
            } catch (RuntimeException e) {
                last = e;
                if (i < attempts) {
                    try { Thread.sleep(delay.toMillis()); } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw e;
                    }
                }
            }
        }
        throw last;
    }
}

File: src/main/kotlin/utils/RetryUtil.kt
package utils

import java.time.Duration

object RetryUtil {
    fun run(attempts: Int, delay: Duration, action: () -> Unit) {
        var last: RuntimeException? = null
        for (i in 1..attempts) {
            try {
                action()
                return
            } catch (e: RuntimeException) {
                last = e
                if (i < attempts) {
                    try {
                        Thread.sleep(delay.toMillis())
                    } catch (ie: InterruptedException) {
                        Thread.currentThread().interrupt()
                        throw e
                    }
                }
            }
        }
        throw last ?: RuntimeException("Unknown error in retry")
    }
}

